SCP-737 Emulator Refactoring

j-mueller opened the pull request:
This PR changes the dynamic behaviour of the emulator. The emulated components (node, wallet, chain index, etc) are still the same. What is new is how we get from "the user defines a sequence of actions" to "operations on the emulated components are performed in the right order".

# The problem

Emulator traces currently require the user to put in 'handleBlockchainEvents' statements in random places. This has confused everyone who has tried to use the emulator (apart from me) in the past, because the statement really depends on the internal state of the contract instance, which should be hidden from the user. 

# Solution

Putting 'handleBlockchainEvents' in random places is necessary because contract instances and the node are separate processes that run in parallel *in the real world*, whereas in the emulator they are all part of the same process. So to solve this problem I added a scheduler that runs the contract instances and the emulated blockchain in separate "threads" (in the end, everything still uses a single GHC thread. See the documentation for Plutus.Trace.Scheduler for details)

A second big change is to the results produced by the emulator: It now produces a stream of events. Previously we just had the final state to look at. 

# Outcomes

* No more 'handleBlockchainEvents'
* Better logging
	- including logs generated by contract instances. See for example crowdfundingEmulatorTestOutput.txt
	- all logs are totally ordered and annotated with a timestamp in simulator time (Slots)
* Being able to run several different contracts in the emulator
	- Previously this was just a single contract with one instance for each of the wallets
* It is now easier to look at the dynamic changes of the system over time thanks to the concept of the emulator event stream
	- See 'Wallet.Emulator.Folds'
* Improved support for property tests
	- trace predicates can be used in unit tests and in property tests
	- See the tests in 'Spec.Emulator'

# Docs

I added a whole bunch of notes explaining the rationale behind some of the decisions:

* Scheduler
	[Thread Tag]
	[Thread Priority]
	[Freeze and Thaw]
* Emulator system
	[Wallet contract instances]
	[Simulator Time]
	[Simulated Agents]
* Traces
	[Trace]
	[Playground traces]
	[The EmulatorControl effect]
	[Emulator event stream]

j-mueller commented:
@michaelpj  This is what have, let's talk about it in the call tomorrow

j-mueller commented:
I added a new version of the scheduler. The biggest change is that the threads now look like this: `newtype Worker effs = Worker {unWorker :: InboxMsg -> Eff effs [OutboxMsg]}`, ie. no more continuations. Not sure what to think about it yet.

michaelpj commented on plutus-contract/src/Plutus/Trace/Types.hs:
```diff
@@ -0,0 +1,269 @@
+loop2 = do
+  currentState <- get @(MBState effs)
+  unless (HashMap.null $ mbInboxes currentState) $ do
+    State.gets @(MBState effs) (HashMap.keysSet . view inboxes) >>= traverse_ handleWorker
```
`handleWorker` will throw when it has no messages, rather than skipping, so I don't think this will work right.

j-mueller replied:
The map only contains non-empty sequences (when a worker has no more messages then the entry is deleted from the map)

This is so that we can easily check if all the threads are done, via `HashMap.null`

j-mueller replied:
I should probably look for non-empty sequence/queue implementation

michaelpj replied:
Ah right, you only traverse over the threads with non-empty messageboxes. Could be simpler, but not a big deal.

j-mueller commented:
# Summary of discussion & current status

NB I'm going to use the word thread for "actions that run interleaved ('in parallel')" - this does not mean GHC RTS threads, or OS threads.

## Concerns about writing our own concurrency / why do we need it

* The emulator is emulating a concurrent system, and the concurrency of the emulated system is likely to be reflected *somewhere* in the emulator
* At the moment, this place is in the actual traces.
	* We need to insert `handleBlockchainEvents` and `addSlot` statements in the correct sequence.
		* For example, in a test for the RPC feature we have three `handleBlockchainEvents`, then one `addSlot`, then another `handleBlockchainEvents`.
	* So our traces are like manually written schedules detailing when each thread should run
	* This is hard to get right (ask Alex), and any change in the definition of the contracts we are emulating is likely to require a change in the thread schedule
* The solution is to replace the manual schedule with an automatic scheduler.

### What should our threads be able to wait on?

This depends on the kind of thread.

* Contract instances.
	* Slot changes, rollbacks: May happen at any time, initiated by node
		* In emulated mode, we want to create new blocks in the mockchain, and to advance the clock, as soon as we can (faster than real-time).
		* In interactive mode the slot/block/rollback events come from the real node, driven by an external clock
	* User input: May happen at any time, initiated by user
		* In emulated mode, all of the user inputs are provided as part of the trace, with instructions detailing when the input should be provided to the instance (for example, when the endpoint is available, or after X slots)
		* In interactive mode, the user inputs can come from the actual UI, or from the trace
	* Wait for other threads to make progress (Example: RPC)
* "Networking". (Either the real node/network, or the emulated node/network)
	* Doesn't wait on anything, constantly churns out new blocks & rollbacks
* User
	* Wait for endpoints on contract instances to become available (more generally, wait for instances to enter a specific state)
	* Potentially also wait for slot changes?

So we have two different levels of sleeping that we need to consider in our scheduler:

*Level 1*. Yield to other running threads, then resume
*Level 2*. Resume only when an external event happens (user input/slot change/rollback)

### We also need some system calls.

1. Contract instances
	* Send messages to other contract instances (RPC)
2. User
	* Start contract instances. This should return a threadID which can then pass to another contract instance as user input (to enable RPC)

## Other concerns

> What happens if a user input or slot change event comes in and the thread is busy doing something else? Should it be interrupted?

In both approaches, there is a step where we evaluate some function `a -> m b`, and while that happens we do not process any other events. We can't realistically interrupt the evaluation, so we have to wait until the `b` is computed before we process the event. So any time the scheduler gets to look at a system events, *all threads are in a suspended state*.

And we can always tell what a suspended thread is waiting for.
* If it's waiting for the event that came in then we resume the thread
* If it's waiting for other threads to make progress then the arrival of the event doesn't change that particular thread and it can stay suspended until other threads have actually made progress.

# Description of the two approaches.

Both approaches rely on queues that the scheduler cycles through to advance the running threads. The difference is in what's in the queues.

## Continuation based approach

In the continuation approach the queues contain continuations.
* There are two queues, one for each level of sleeping (see above)
* The scheduler cycles through the list of level-1 threads. Each of those thread either finishes completely (is done), or it goes back to sleep at level 1 or level 2. In that case the thread is placed in the corresponding queue.
* When the list of level-1 threads is empty, the scheduler waits for a system event to happen (by calling the somewhat unfortunately named `interpOnSlotChange` handler)
* The scheduler then moves all level-2 threads to the level-1 queue and starts again.

## Inbox based approach

In the inbox approach, there is one queue per running thread - the thread's inbox.

* Threads are functions `inboxMessage -> m [outboxMessage]` (no continuations). 
* The scheduler looks at each thread's inbox.
* If the inbox contains a message, it runs the thread on the message.
* Then it takes the outbox messages the thread produced and routes them the inboxes of their recipients.
* Then it starts again until all inboxes are empty

# Differences between the approaches

## Handling system events

The main difference is how system events are handled.

* In the continuation based approach, the scheduler calls `interpOnSlotChange` to get the next system event
* In the inbox based approach, system events are generated by the threads themselves. There is no way for the scheduler to ask for the next system event

### Handling system events in interactive mode

* In the continuation based approach, `interpOnSlotChange` is a blocking read from a TVar, and the UI and node client write to that TVar
* In the inbox based approach,  one of the running threads is a (blocking?) read from a TVar, returning a list of outbox messages for all the threads that are waiting for system events
	- How does it know which threads are waiting for system events? It probably needs to ask the scheduler. How does the scheduler know?
	
### Handling system events in emulated mode

* In the continuation based approach we define `interpOnSlotChange` to make a new block in the mockchain and advance the internal clock
* In the inbox based approach: ?? How do we know when to make a new block?

## Interpreting traces

How do we translate a trace to a thread (or list of threads)?

* Continuation based approach: The trace interpreter is itself a thread, using the system calls (fork/wait for slot/etc) to drive the trace.
	- It forks one thread per contract instance
	- Calling endpoints is handled in the contract instance thread
		- In emulated mode, when we encounter a "call endpoint" instruction we can wait until the endpoint is available on the instance
		- In interactive mode we can read the input from an instance-specific TVar
* Inbox based approach:
	- Calling endpoints is handled in the interpreter thread
		- How does it know when to produce the "call endpoint" messages?
		- In interactive mode we can read from an instance-specific TVar

# Discussion

* The inbox based approach seems a bit clunky.
	* We need to pack all requests and responses into a big union type, which includes both the system calls and the system events
	* In emulated mode, I'm not sure how to get the "trace interpreter" thread to produce the "next slot" and "user input" events at the correct time
		* It should only produce them when all other threads have empty inboxes
		* How does it know that? It needs some kind of special privilege to look at the inboxes or query the scheduler for this information
		* But then it's not a normal thread anymore!
* The continuation based approach could be harder to debug if something goes wrong
	* But even here the emulator still only runs one thread at a time, so we don't have any real concurrency (although we could add it later on)
	* In both approaches a lot of the state of the system is going to be captured in some `State` effect, and we'd have to look at that state for debugging
* The continuation based approach integrates nicely with the other `freer-simple` effects, due to it just being an instance of the `Yield` effect
	* As a result, we can write system calls as just another effect that then gets interpreted in terms of `Yield`. This effect is available to the threads themselves.

j-mueller commented:
@michaelpj There are a number of timing issues that I'm unclear on in the message based approach. See above. (It's quite a bit of text, but there are a lot of things to consider)

j-mueller commented:
I think the underlying problem is that the threads in the inbox approach don't have any state, at least the trace interpreter thread really needs its own state to keep track of the trace (which instructions have been processed already and which ones are yet to come)

michaelpj commented:
> In the inbox based approach, one of the running threads is a (blocking?) read from a TVar, returning a list of outbox messages for all the threads that are waiting for system events 

This isn't quite what I had in mind. Rather, one of the threads might read from a TVar containing user input, and then *sends* messages to all the threads regardless of whether they're waiting for a system event (we assume that a thread can handle a system event it doesn't care about very quickly, so it isn't a problem to broadcast them).

> In the inbox based approach: ?? How do we know when to make a new block?

The "network" thread would have to decide somehow. It depends on what policy we want (e.g. wait until no contracts are doing anything).

> How do we translate a trace to a thread (or list of threads)?

I'm not sure this quite makes sense in interactive mode: interactive mode *produces* a trace of events that happened in the course of evaluation, each time a message gets sent somewhere. You could, for example, then re-interpret this to replay the events!

> We need to pack all requests and responses into a big union type, which includes both the system calls and the system events

Does the type family stuff help with this? It's really `LocalAction | GlobalAction`, right?

> In emulated mode, I'm not sure how to get the "trace interpreter" thread to produce the "next slot" and "user input" events at the correct time 

This seems tricky, I agree. But this is only a problem because we want to roll it into the next slot/user input action. If we were willing to have a `runUntilSettled` action (which is a bit like `runBlockchainActions`, maybe that's a problem), then we could require people to do that manually (or insert it into the trace in the `nextSlot` function).

michaelpj commented:
>I think the underlying problem is that the threads in the inbox approach don't have any state, at least the trace interpreter thread really needs its own state to keep track of the trace (which instructions have been processed already and which ones are yet to come)

I'm not quite sure I follow this, in particular I'm not sure I see where the "trace interpreter" thread comes from. In emulated mode we do go through the trace step by step, but that's fine, I think (modulo the question above about when to do the network actions).

In simulated mode there's no "trace interpreter", there's just user input and the network thread throwing actions in at intervals.

j-mueller commented:
> This isn't quite what I had in mind

Ok thanks for the clarification.



> > We need to pack all requests and responses into a big union type, which includes both the system calls and the system events

> Does the type family stuff help with this? It's really LocalAction | GlobalAction, right?

On second thought, I think the problem isn't that big because there are only two different system events (user input, slot change). I had it confused with the contract-specific messages such as `PubKeyRequest` but we can handle them using the `effs` effects without using the thread system. So I retract this objection.

> You could, for example, then re-interpret this to replay the events!

Yes that's a very good argument for the reified messages.

> If we were willing to have a `runUntilSettled` action (which is a bit like runBlockchainActions, maybe that's a problem)

In my view this is indeed a problem, it shouldn't be necessary because in theory we have all the information needed to decide when to make a new block in emulated mode.

> I'm not quite sure I follow this, in particular I'm not sure I see where the "trace interpreter" thread comes from.

I was thinking of [`handleEmulator`](https://github.com/input-output-hk/plutus/blob/7ef79f5007cee6e23e45e4fe8e4fc90c7ed254aa/plutus-contract/src/Plutus/Trace/Types.hs#L98-L111), which I agree isn't a thread, although it does use the `Yield` effect.

Maybe there is a way to reconcile the continuation/``Yield` idea with the mailboxes, to get the replayability you described.

j-mueller commented:
I changed the `NextSlot` handler to

```haskell
 newtype FetchSystemEvent effs systemEvent = FetchSystemEvent {runFetchSystemEvent :: Eff effs systemEvent } 
```

I think that's a better description of what it actually does.

j-mueller commented:
Hm, what if we use `FetchSystemEvent` in the mailbox scheduler? Then the scheduler would fetch the next event whenever all mailboxes are empty. So we get the inboxes but no more need for `handleBlockchainEvents`

michaelpj commented:
> Hm, what if we use FetchSystemEvent in the mailbox scheduler?

Is the "mailbox scheduler" the thread that's interpreting the trace? I think it would make sense for it to emit the "network" events when we're in emulated mode.

Here's an argument *against* the automatic block creation. If we take seriously the idea that interactive mode could *produce* a trace that we can later rerun, then that trace will include the network events when they happened in the real log. We don't then want to insert more block creations! So there's some use for a "strict" emulator mode that really just does exactly what it's told, and has to be told when to put blocks in.

I wonder if we can build the "automatic block creation" functionality on top of the "strict" version - maybe that would fit together better?

j-mueller commented:
> I wonder if we can build the "automatic block creation" functionality on top of the "strict" version - maybe that would fit together better?

I think it could work the other way around. We could change the signature of `runFetchSystemEvent` to return `Maybe systemEvent`, and then the scheduler could stop as soon as it receives a `Nothing`. Then we define `runFetchSystemEvent = pure Nothing` so that it stops every time all threads have settled.

j-mueller commented:
I changed the mailbox scheduler to use `FetchSystemEvent`, what do you think about it?

Also, I think if we don't have stateful threads we will end up with a `State (Map ThreadId ThreadState)` somewhere in our `effs`, which isn't very nice *and* requires the threads' states to be of the same type. So I changed the worker type to

```haskell
newtype Worker e effs = Worker {unWorker :: InboxMsg e -> Eff effs ([OutboxMsg e], Maybe (Worker e effs)) }
```

j-mueller commented:
Also, I realised my earlier statement

> I was thinking of handleEmulator, which I agree isn't a thread, although it does use the Yield effect.

While not incorrect, was talking about the wrong function. If you look at [`runThreads`](https://github.com/input-output-hk/plutus/pull/2314/files#diff-e2001a7c24351c1583fb043bbbdb9493R147-R156), the argument `e` is exactly the trace interpreter, and it does get scheduled like a normal thread. Which I think is very nice.

michaelpj commented:
I don't understand why we would need a map of `ThreadState`s. A thread may be stateful, yes, but can't that just live within the thread? e.g. if we have a thread that represents one participant, i.e. is effectively a PAB, then sure, it'll need to keep the suspended contract instance so it can advance it when it gets more messages. But at least in the actual-thread version it seems like we should be able to keep this internally, i.e. the thread can be doing 
```
go :: State -> m ()
go s = do
    e <- pollEvent
    s' <- doStuff s
    go s
```

Maybe this is more complicated in the "not-real-threads" version, but I feel like it should be able to work.

michaelpj commented:
The mailboxes need to be "shared state" because they are shared between the communicating parties, the internal thread state shouldn't be.

j-mueller commented:
Yes that is exactly what I am proposing, that's why I think the `Worker` type should be

```haskell
newtype Worker e effs = Worker {unWorker :: InboxMsg e -> Eff effs ([OutboxMsg e], Maybe (Worker e effs)) }
```

instead of

```haskell
newtype Worker e effs = Worker {unWorker :: InboxMsg e -> Eff effs [OutboxMsg e] }
```

which it used to be, and which does not allow you to store the state internally.

michaelpj commented:
> Yes that is exactly what I am proposing (for the mailbox version), that's why I think the Worker type should be

But then we've got the continuation in there again! So we're just stuffing everything into that. Why can't we use our `effs` to do the local state?

j-mueller commented:
> Why can't we use our effs to do the local state?

All threads use the same `effs`, so I don't understand where the local state would go in `InboxMsg e -> Eff effs [OutboxMsg e]`, unless it's part of `effs`, in which case it is going to be a `Map ThreadId ThreadState`.

j-mueller commented:
From the call today:

* MPJ: Perturbing the system is what we should be recording
  * In that case the continuations are not so bad
  * We want to be able to test race conditions
  * Anything the scheduler *could* do non-deterministically in interactive mode we should able to model in deterministic mode using randomness, for example randomising the queue order

* JM: Mailboxes are necessary

* System messages thread could be the lowest priority thread
	* In interpreted mode, what if we had no system thread, but instead the trace says:
	    - Wait until no more requests
	    - Wait until they have turned off all their endpoints
	    - Wait until a particular contract has terminated (try not to get stuck)
	    - Next slot (via trace, timed thread, user action)
	    - Run until someone does something (one of the contracts changes its state)
	* The script has to be able to inspect the threads (using an effect that is handled by the emulator)
	* The trace is a specification of what system events happend, and some user inputs at specific times

j-mueller commented on plutus-contract/src/Control/Monad/Freer/Extras.hs:
```diff
@@ -229,6 +234,12 @@ raiseEnd17 = loop where
         Val a -> pure a
         E u q -> E (weakenEnd17 u) (tsingleton $ qComp q loop)
 
+raiseEnd18 :: forall effs a b c d e f g h i j k l m n o p q r. Eff '[a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r] ~> Eff (a ': b ': c ': d ': e ': f ': g ': h ': i ': j ': k ': l ': m ': n ': o ': p ': q ': r ': effs)
```
:'(

michaelpj replied:
:scream_cat:

j-mueller commented:
@michaelpj  I think this is in a good state to be reviewed now. Compared to when we last talked about it I added the `Wallet.Emulator.Stream` and `Wallet.Emulator.Folds` modules, split up the `Emulator` effect into several smaller ones and changed everything that uses the emulator to use the new structure,

Still left to do is updating the PAB. It currently has its own version of the emulator. There is a separate Jira issue for that, and I didn't want to make the PR even bigger than it is already.

j-mueller commented:
A good place to start the review is probably `Plutus.Trace`

michaelpj commented:
@gilligan I added you as a reviewer on this. I'm not really expecting you to necessarily have much to say, but perhaps a good opportunity to stare at some more of the innards and ask questions!

michaelpj commented:
Okay, I think at this point we should just merge this and move forwards!
